In [52]:
! pip install mysql-connector-python


[notice] A new release of pip is available: 23.2.1 -> 25.0
[notice] To update, run: pip install --upgrade pip


## Importing Libraries ##
 

About this file

Suggest Edits
This dataset provides comprehensive details about registered electric vehicles (EVs) in various regions, primarily focusing on Washington State. It includes crucial information such as vehicle identification numbers (VINs), location details (county, city, state, postal code), and other metadata. The dataset can be useful for transportation research, policy-making, and sustainability studies.

Dataset Contents
VIN (Vehicle Identification Number) – A unique identifier for each registered electric vehicle.

County – The county where the vehicle is registered.

City – The specific city of registration.

State – The state of registration (mostly WA - Washington).

Postal Code – The zip code for further geographic granularity

In [130]:
import pandas as pd
import sqlite3
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', None)

### Reading data from csv to a dataframe

Data used from Kaggle

In [131]:
name = "ev.db"
table_name = "ev_table_3"
df = pd.read_csv('/Users/vardasinghal/Projects/EV_Analysis/Electric_Vehicle_Population_Data.csv')


### Data Preparation

In [132]:
df.head(2)

VIN (1-10)     County          City State  Postal Code  Model Year  \
0  1C4JJXP66P     Kitsap       Poulsbo    WA      98370.0        2023   
1  1G1FX6S08K  Snohomish  Lake Stevens    WA      98258.0        2019   

        Make     Model                   Electric Vehicle Type  \
0       JEEP  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
1  CHEVROLET   BOLT EV          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0             Not eligible due to low battery range            21.0   
1           Clean Alternative Fuel Vehicle Eligible           238.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  23.0       258127145   
1        0.0                  44.0         4735426   

              Vehicle Location        Electric Utility  2020 Census Tract  
0  POINT (-122.64681 47.73689)  PUGET SOUND ENERGY INC       5.303509e+10  
1  POINT (-122.06402 48.01497)  PUGET SOUND ENERGY INC       5.306105e+10

In [133]:
import pandas as pd

# List of current column names
columns = df.columns

# Create a list of updated column names (replacing spaces and slashes with underscores)
updated_columns = [col.replace(' ', '_').replace('/', '_') for col in columns]

# Assign the updated column names to the DataFrame
df.columns = updated_columns

# Check if the columns have been renamed
print(df.columns)

Index(['VIN_(1-10)', 'County', 'City', 'State', 'Postal_Code', 'Model_Year',
       'Make', 'Model', 'Electric_Vehicle_Type',
       'Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility', 'Electric_Range',
       'Base_MSRP', 'Legislative_District', 'DOL_Vehicle_ID',
       'Vehicle_Location', 'Electric_Utility', '2020_Census_Tract'],
      dtype='object')


In [57]:
df.head(5)

VIN_(1-10)     County          City State  Postal_Code  Model_Year  \
0  1C4JJXP66P     Kitsap       Poulsbo    WA      98370.0        2023   
1  1G1FX6S08K  Snohomish  Lake Stevens    WA      98258.0        2019   
2  WBY1Z2C58F       King       Seattle    WA      98116.0        2015   
3  5YJ3E1EBXK       King       Seattle    WA      98178.0        2019   
4  5YJSA1V24F     Yakima         Selah    WA      98942.0        2015   

        Make     Model                   Electric_Vehicle_Type  \
0       JEEP  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
1  CHEVROLET   BOLT EV          Battery Electric Vehicle (BEV)   
2        BMW        I3          Battery Electric Vehicle (BEV)   
3      TESLA   MODEL 3          Battery Electric Vehicle (BEV)   
4      TESLA   MODEL S          Battery Electric Vehicle (BEV)   

  Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility  Electric_Range  \
0             Not eligible due to low battery range            21.0   
1           Clean Alternative Fuel Vehicle Eligible           238.0   
2           Clean Alternative Fuel Vehicle Eligible            81.0   
3           Clean Alternative Fuel Vehicle Eligible           220.0   
4           Clean Alternative Fuel Vehicle Eligible           208.0   

   Base_MSRP  Legislative_District  DOL_Vehicle_ID  \
0        0.0                  23.0       258127145   
1        0.0                  44.0         4735426   
2        0.0                  34.0       272697666   
3        0.0                  37.0       477309682   
4        0.0                  15.0       258112970   

              Vehicle_Location                              Electric_Utility  \
0  POINT (-122.64681 47.73689)                        PUGET SOUND ENERGY INC   
1  POINT (-122.06402 48.01497)                        PUGET SOUND ENERGY INC   
2  POINT (-122.41067 47.57894)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
3  POINT (-122.23825 47.49461)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
4  POINT (-120.53145 46.65405)                                    PACIFICORP   

   2020_Census_Tract  
0       5.303509e+10  
1       5.306105e+10  
2       5.303301e+10  
3       5.303301e+10  
4       5.307700e+10

### Connecting to database

In [134]:
#connect to a database so that I can use SQL
conn = sqlite3.connect(name) #if the db does not exist, this creates a Any_Database_Name.db file in the current directory
#store your table in the database:
df.to_sql("table_4", conn)

ValueError: Table 'table_4' already exists.

In [135]:
#Top Cities by Registration
get_city_split = ''' 
Select City , Count(City) as city_split
from table_4
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
''' 
get_city_split_df = pd.read_sql(get_city_split, conn)
get_city_split_df


City  city_split
0    Seattle       35664
1   Bellevue       10966
2  Vancouver        8103
3    Redmond        7772
4    Bothell        7298
5     Renton        6463
6   Kirkland        6438
7  Sammamish        6290
8    Olympia        5419
9     Tacoma        4818

In [136]:
df.head(3)

VIN_(1-10)     County          City State  Postal_Code  Model_Year  \
0  1C4JJXP66P     Kitsap       Poulsbo    WA      98370.0        2023   
1  1G1FX6S08K  Snohomish  Lake Stevens    WA      98258.0        2019   
2  WBY1Z2C58F       King       Seattle    WA      98116.0        2015   

        Make     Model                   Electric_Vehicle_Type  \
0       JEEP  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
1  CHEVROLET   BOLT EV          Battery Electric Vehicle (BEV)   
2        BMW        I3          Battery Electric Vehicle (BEV)   

  Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility  Electric_Range  \
0             Not eligible due to low battery range            21.0   
1           Clean Alternative Fuel Vehicle Eligible           238.0   
2           Clean Alternative Fuel Vehicle Eligible            81.0   

   Base_MSRP  Legislative_District  DOL_Vehicle_ID  \
0        0.0                  23.0       258127145   
1        0.0                  44.0         4735426   
2        0.0                  34.0       272697666   

              Vehicle_Location                              Electric_Utility  \
0  POINT (-122.64681 47.73689)                        PUGET SOUND ENERGY INC   
1  POINT (-122.06402 48.01497)                        PUGET SOUND ENERGY INC   
2  POINT (-122.41067 47.57894)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   

   2020_Census_Tract  
0       5.303509e+10  
1       5.306105e+10  
2       5.303301e+10

In [137]:
#EV Adoption over years
get_ev_adoption_split = ''' 
Select model_year , Count(*) as ev_adoption
from table_4
GROUP BY 1
ORDER BY 2 DESC
''' 
get_ev_adoption_split_df = pd.read_sql(get_ev_adoption_split, conn)
get_ev_adoption_split_df

Model_Year  ev_adoption
0         2023        59060
1         2024        45525
2         2022        28610
3         2021        20278
4         2018        14142
5         2020        12103
6         2019        10874
7         2017         8433
8         2016         5358
9         2025         5007
10        2015         4680
11        2013         4258
12        2014         3404
13        2012         1513
14        2011          692
15        2010           23
16        2008           23
17        2000            7
18        2002            2
19        1999            2
20        2003            1

In [138]:
#Ev Split type Split
get_ev_type_split = ''' 
Select Electric_Vehicle_Type , Count(Electric_Vehicle_Type) as ev_split
from table_4
GROUP BY 1
ORDER BY 2 DESC
''' 
get_ev_type_split_df = pd.read_sql(get_ev_type_split, conn)
get_ev_type_split_df

Electric_Vehicle_Type  ev_split
0          Battery Electric Vehicle (BEV)    177151
1  Plug-in Hybrid Electric Vehicle (PHEV)     46844

In [143]:
#Top 10 makes by Vehicle type
get_leaders_type_split = ''' 
WITH CTE AS
(
Select  Make , Electric_Vehicle_Type , Count(Electric_Vehicle_Type) as ev_split,
DENSE_RANK() OVER(PARTITION BY  Electric_Vehicle_Type ORDER BY Count(Electric_Vehicle_Type) DESC ) as rnk
from table_4
GROUP BY 1 , 2
ORDER BY 3 DESC
)
select Make , Electric_Vehicle_Type , ev_split
from  cte
where rnk <= 10
order by 2
''' 
get_leaders_type_split_df = pd.read_sql(get_leaders_type_split, conn)
get_leaders_type_split_df

Make                   Electric_Vehicle_Type  ev_split
0        TESLA          Battery Electric Vehicle (BEV)     96180
1       NISSAN          Battery Electric Vehicle (BEV)     15259
2    CHEVROLET          Battery Electric Vehicle (BEV)     11685
3         FORD          Battery Electric Vehicle (BEV)      8234
4          KIA          Battery Electric Vehicle (BEV)      7688
5       RIVIAN          Battery Electric Vehicle (BEV)      6236
6      HYUNDAI          Battery Electric Vehicle (BEV)      5883
7   VOLKSWAGEN          Battery Electric Vehicle (BEV)      5740
8          BMW          Battery Electric Vehicle (BEV)      3604
9         AUDI          Battery Electric Vehicle (BEV)      2375
10      TOYOTA  Plug-in Hybrid Electric Vehicle (PHEV)      7730
11        JEEP  Plug-in Hybrid Electric Vehicle (PHEV)      5773
12         BMW  Plug-in Hybrid Electric Vehicle (PHEV)      5567
13   CHEVROLET  Plug-in Hybrid Electric Vehicle (PHEV)      4720
14       VOLVO  Plug-in Hybrid Electric Vehicle (PHEV)      3751
15    CHRYSLER  Plug-in Hybrid Electric Vehicle (PHEV)      3733
16        FORD  Plug-in Hybrid Electric Vehicle (PHEV)      3696
17         KIA  Plug-in Hybrid Electric Vehicle (PHEV)      3072
18        AUDI  Plug-in Hybrid Electric Vehicle (PHEV)      1782
19  MITSUBISHI  Plug-in Hybrid Electric Vehicle (PHEV)      1013

In [42]:
#Split by make
get_make_split = ''' 
Select make , Count(make) as make_split
from table_4
GROUP BY 1
ORDER BY 2 DESC
limit 10
''' 
get_make_split_df = pd.read_sql(get_make_split, conn)
get_make_split_df

Make  make_split
0      TESLA       96180
1  CHEVROLET       16405
2     NISSAN       15259
3       FORD       11930
4        KIA       10760
5        BMW        9171
6     TOYOTA        8956
7    HYUNDAI        6878
8     RIVIAN        6236
9       JEEP        5773

In [97]:
#Top 10 EV vehicle registrations
get_model_split = ''' 
Select make,  model,  Count(model) as model_split
from table_4
GROUP BY 1 , 2
ORDER BY 3 DESC
LIMIT 10
''' 
get_model_split_df = pd.read_sql(get_model_split, conn)
get_model_split_df


Make           Model  model_split
0       TESLA         MODEL Y        46583
1       TESLA         MODEL 3        34462
2      NISSAN            LEAF        13735
3       TESLA         MODEL S         7765
4   CHEVROLET         BOLT EV         7008
5       TESLA         MODEL X         6243
6        FORD  MUSTANG MACH-E         4774
7   CHEVROLET            VOLT         4720
8  VOLKSWAGEN            ID.4         4653
9        JEEP        WRANGLER         4292

In [86]:
#Top Vehicle registrations for each make by make ,  model and  year
get_year_split = ''' 
WITH CTE AS (
    SELECT make, model, model_year,  COUNT(*) AS registrations,
           DENSE_RANK() OVER(PARTITION BY make ORDER BY COUNT(*) DESC) AS rnk
    FROM table_4
    GROUP BY make, model , model_year
)
SELECT make, model, model_year, registrations
FROM CTE
WHERE rnk = 1
ORDER BY registrations DESC
''' 
get_year_split_df = pd.read_sql(get_year_split, conn)
get_year_split_df


make                     model  model_year  \
0                    TESLA                   MODEL Y        2023   
1               VOLKSWAGEN                      ID.4        2023   
2                     FORD            MUSTANG MACH-E        2023   
3                CHEVROLET                  BOLT EUV        2023   
4                  HYUNDAI                   IONIQ 5        2023   
5                   NISSAN                      LEAF        2013   
6                     JEEP                  WRANGLER        2024   
7                   RIVIAN                       R1S        2023   
8                   TOYOTA                RAV4 PRIME        2024   
9                 CHRYSLER                  PACIFICA        2023   
10                     KIA                       EV9        2024   
11                   HONDA                  PROLOGUE        2024   
12                  SUBARU                  SOLTERRA        2024   
13                CADILLAC                     LYRIQ        2024   
14                     BMW                        I4        2024   
15                   MAZDA                     CX-90        2024   
16                   DODGE                    HORNET        2024   
17                   VOLVO                      XC40        2023   
18                    AUDI                    E-TRON        2019   
19                POLESTAR                       PS2        2022   
20              MITSUBISHI                 OUTLANDER        2023   
21           MERCEDES-BENZ                 EQB-CLASS        2023   
22                    MINI                   HARDTOP        2024   
23                    FIAT                       500        2015   
24                   LEXUS                        RZ        2023   
25                     GMC             HUMMER EV SUV        2024   
26                 PORSCHE                    TAYCAN        2022   
27                  FISKER                     OCEAN        2023   
28                   LUCID                       AIR        2023   
29                 GENESIS                      GV60        2023   
30                   ACURA                       ZDX        2024   
31                  JAGUAR                    I-PACE        2019   
32              ALFA ROMEO                    TONALE        2024   
33                 LINCOLN                   CORSAIR        2022   
34                   SMART     FORTWO ELECTRIC DRIVE        2014   
35              LAND ROVER         RANGE ROVER SPORT        2020   
36                   TH!NK                      CITY        2011   
37          AZURE DYNAMICS  TRANSIT CONNECT ELECTRIC        2011   
38             ROLLS-ROYCE                   SPECTRE        2024   
39    WHEEGO ELECTRIC CARS                    WHEEGO        2010   
40              BRIGHTDROP                      ZEVO        2023   
41             LAMBORGHINI                  REVUELTO        2024   
42  MULLEN AUTOMOTIVE INC.                       ONE        2024   
43                     RAM            PROMASTER 3500        2024   
44                 BENTLEY                  BENTAYGA        2021   
45                 BENTLEY                  BENTAYGA        2023   
46                 BENTLEY                  BENTAYGA        2024   
47                 BENTLEY               FLYING SPUR        2022   
48                 VINFAST                      VF 8        2023   
49                 VINFAST                      VF 8        2024   

    registrations  
0           17998  
1            2451  
2            2097  
3            2066  
4            1853  
5            1793  
6            1693  
7            1611  
8            1515  
9            1310  
10           1204  
11            932  
12            914  
13            850  
14            759  
15            758  
16            733  
17            539  
18            515  
19            387  
20            358  
21            324  
22            243  
23            222  
24            201  
25            191  
26           

In [70]:
#Split by Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility by make and model
get_CAFV_cnt_split = ''' 
Select "Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility", COUNT("Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility") as CAFV_cnt_split
from table_4
GROUP BY 1
ORDER BY 2 DESC
''' 
get_CAFV_cnt_split_df = pd.read_sql(get_CAFV_cnt_split, conn)
get_CAFV_cnt_split_df

Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility  \
0  Eligibility unknown as battery range has not been researched   
1                       Clean Alternative Fuel Vehicle Eligible   
2                         Not eligible due to low battery range   

   CAFV_cnt_split  
0          130442  
1           71438  
2           22115

In [77]:
#Split by Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility by make and model
get_CAFV_split = ''' 
Select make , model , "Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility", COUNT("Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility") as CAFV_split
from table_4
GROUP BY 1
ORDER BY 4 DESC
LIMIT 20 
''' 
get_CAFV_split_df = pd.read_sql(get_CAFV_split, conn)
get_CAFV_split_df

Make        Model  \
0           TESLA      MODEL 3   
1       CHEVROLET      BOLT EV   
2          NISSAN         LEAF   
3            FORD        F-150   
4             KIA          EV6   
5             BMW           I3   
6          TOYOTA  PRIUS PRIME   
7         HYUNDAI     SANTA FE   
8          RIVIAN          R1T   
9            JEEP     WRANGLER   
10     VOLKSWAGEN         ID.4   
11          VOLVO         XC60   
12           AUDI       E-TRON   
13       CHRYSLER     PACIFICA   
14  MERCEDES-BENZ    GLC-CLASS   
15         SUBARU     SOLTERRA   
16          HONDA      CLARITY   
17        PORSCHE      CAYENNE   
18       POLESTAR          PS2   
19           MINI      HARDTOP   

               Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility  CAFV_split  
0                        Clean Alternative Fuel Vehicle Eligible       96180  
1                        Clean Alternative Fuel Vehicle Eligible       16405  
2                        Clean Alternative Fuel Vehicle Eligible       15259  
3   Eligibility unknown as battery range has not been researched       11930  
4   Eligibility unknown as battery range has not been researched       10760  
5                        Clean Alternative Fuel Vehicle Eligible        9171  
6                          Not eligible due to low battery range        8956  
7                        Clean Alternative Fuel Vehicle Eligible        6878  
8   Eligibility unknown as battery range has not been researched        6236  
9                          Not eligible due to low battery range        5773  
10  Eligibility unknown as battery range has not been researched        5740  
11                       Clean Alternative Fuel Vehicle Eligible        5565  
12  Eligibility unknown as battery range has not been researched        4157  
13                       Clean Alternative Fuel Vehicle Eligible        3733  
14                         Not eligible due to low battery range        2239  
15  Eligibility unknown as battery range has not been researched        1830  
16                       Clean Alternative Fuel Vehicle Eligible        1802  
17                         Not eligible due to low battery range        1377  
18                       Clean Alternative Fuel Vehicle Eligible        1222  
19  Eligibility unknown as battery range has not been researched        1085

In [80]:
df.head()

VIN_(1-10)     County          City State  Postal_Code  Model_Year  \
0  1C4JJXP66P     Kitsap       Poulsbo    WA      98370.0        2023   
1  1G1FX6S08K  Snohomish  Lake Stevens    WA      98258.0        2019   
2  WBY1Z2C58F       King       Seattle    WA      98116.0        2015   
3  5YJ3E1EBXK       King       Seattle    WA      98178.0        2019   
4  5YJSA1V24F     Yakima         Selah    WA      98942.0        2015   

        Make     Model                   Electric_Vehicle_Type  \
0       JEEP  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
1  CHEVROLET   BOLT EV          Battery Electric Vehicle (BEV)   
2        BMW        I3          Battery Electric Vehicle (BEV)   
3      TESLA   MODEL 3          Battery Electric Vehicle (BEV)   
4      TESLA   MODEL S          Battery Electric Vehicle (BEV)   

  Clean_Alternative_Fuel_Vehicle_(CAFV)_Eligibility  Electric_Range  \
0             Not eligible due to low battery range            21.0   
1           Clean Alternative Fuel Vehicle Eligible           238.0   
2           Clean Alternative Fuel Vehicle Eligible            81.0   
3           Clean Alternative Fuel Vehicle Eligible           220.0   
4           Clean Alternative Fuel Vehicle Eligible           208.0   

   Base_MSRP  Legislative_District  DOL_Vehicle_ID  \
0        0.0                  23.0       258127145   
1        0.0                  44.0         4735426   
2        0.0                  34.0       272697666   
3        0.0                  37.0       477309682   
4        0.0                  15.0       258112970   

              Vehicle_Location                              Electric_Utility  \
0  POINT (-122.64681 47.73689)                        PUGET SOUND ENERGY INC   
1  POINT (-122.06402 48.01497)                        PUGET SOUND ENERGY INC   
2  POINT (-122.41067 47.57894)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
3  POINT (-122.23825 47.49461)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
4  POINT (-120.53145 46.65405)                                    PACIFICORP   

   2020_Census_Tract  
0       5.303509e+10  
1       5.306105e+10  
2       5.303301e+10  
3       5.303301e+10  
4       5.307700e+10

In [91]:
#Average Electric Range by make
get_avg_range_split = ''' 
Select make, ROUND(Avg(Electric_Range),2) as average_electric_Range
from table_4
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 10
''' 
get_avg_range_split_df = pd.read_sql(get_avg_range_split, conn)
get_avg_range_split_df

Make  average_electric_Range
0                JAGUAR                  196.84
1  WHEEGO ELECTRIC CARS                  100.00
2                 TH!NK                  100.00
3             CHEVROLET                   85.35
4                  FIAT                   80.63
5                NISSAN                   71.54
6                 TESLA                   62.51
7                 SMART                   61.76
8        AZURE DYNAMICS                   56.00
9               PORSCHE                   51.19

In [93]:
# TOP Electric Range by make and model

get_top_range_split = ''' 
Select make,  model, Electric_Range as top_electric_Range
from table_4
GROUP BY 1, 2
ORDER BY 3 DESC
LIMIT 10
''' 
get_top_range_split_df = pd.read_sql(get_top_range_split, conn)
get_top_range_split_df

Make     Model  top_electric_Range_
0    PORSCHE     MACAN                308.0
1    HYUNDAI      KONA                258.0
2      TESLA  ROADSTER                245.0
3  CHEVROLET   BOLT EV                238.0
4     JAGUAR    I-PACE                234.0
5   POLESTAR       PS2                233.0
6      TESLA   MODEL 3                220.0
7      TESLA   MODEL S                208.0
8      TESLA   MODEL X                200.0
9     NISSAN      LEAF                107.0

In [113]:
#Check Missing values
get_miss_df = '''
SELECT COUNT(*) AS missing_prices
FROM table_4
WHERE Base_MSRP IS NULL OR Base_MSRP = 0
'''

get_miss_split_df = pd.read_sql(get_miss_df, conn)
get_miss_split_df


missing_prices
0          220735

In [139]:
#Group by Make: Average Price & Range by Manufacturer
get_price_range_split = ''' 
SELECT Make, 
       ROUND(AVG(Base_MSRP), 2) AS avg_price, 
       ROUND(AVG(Electric_Range), 2) AS avg_range
FROM table_4
WHERE Base_MSRP > 0
GROUP BY Make
ORDER BY avg_range DESC
''' 
get_price_range_split_df = pd.read_sql(get_price_range_split, conn)
get_price_range_split_df

Make  avg_price  avg_range
0                  TESLA   70252.73     213.49
1   WHEEGO ELECTRIC CARS   32995.00     100.00
2                    KIA   32270.03      95.23
3                 FISKER  102000.00      33.00
4               CHRYSLER   39995.00      32.00
5               CADILLAC   75095.00      31.00
6                  VOLVO   56365.36      17.58
7                 SUBARU   34995.00      17.00
8                    BMW   52890.94      14.56
9                PORSCHE  135391.43      13.94
10                  MINI   36865.82      12.00

## Key Insights from the Electric Vehicle (EV) Analysis:
1) Top Cities by EV Registrations:
Seattle leads with 35,664 registrations, followed by Bellevue (10,966) and Vancouver (8,103). Urban centers and tech hubs dominate EV adoption.

2) EV Adoption Over Time:
2023 saw the highest registrations (59,060), indicating a recent surge in EV adoption. Data for 2024–2025 (future years) may reflect pre-registrations or data entry errors.

3) EV Type Split:
Battery Electric Vehicles (BEVs) dominate (177,151), while Plug-in Hybrids (PHEVs) account for only 46,844 registrations. This aligns with global trends favoring fully electric vehicles. Within BEV, Tesla is the leader and within PHEV, Toyota is leading

4) Vechicle registrations by each make:
Analyzed the model most registered for each make along with the year to see the most bought model within each make

5) Manufacturer Dominance:
Tesla leads with 96,180 registrations, followed by Chevrolet (16,405) and Nissan (15,259). Tesla’s Model Y (46,583) and Model 3 (34,462) are the most popular models.

6) CAFV Eligibility:
130,442 entries have unknown eligibility due to unresearched battery ranges. Only 71,438 are confirmed eligible, highlighting potential gaps in data quality.

7) Electric Range Analysis:
Jaguar has the highest average range (196.84 miles), while Tesla’s average (62.51 miles) seems unusually low, possibly due to older models or data inconsistencies.
The Porsche Macan has the highest single-model range (308 miles).

### Recommendations:

1) Focus on BEVs: Policymakers and manufacturers should prioritize BEV infrastructure given their dominance.

2) Regional Analysis: Investigate why cities like Seattle and Bellevue lead in adoption (e.g., incentives, charging networks).

3) Model-Specific Trends: Explore why hybrids like the Toyota Prius Prime have low eligibility, possibly due to battery limitations.